modules/fedimint-meta-server/src/lib.rs

```diff
@@ -218,24 +218,45 @@ impl Meta {
         let mut to_delete: Vec<MetaKey> = vec![];
         let mut to_submit = vec![];
 
-        for (key, MetaDesiredValue { value, salt }) in desired {
-            if Self::get_submission(dbtx, key, self.our_peer_id)
-                .await
-                .as_ref()
+        for (
+            key,
+            MetaDesiredValue {
+                value: desired_value,
+                salt,
+            },
+        ) in desired
+        {
+            let consensus_value = &Self::get_consensus(dbtx, key).await;
+            let consensus_submission_value =
+                Self::get_submission(dbtx, key, self.our_peer_id).await;
+            if consensus_submission_value.as_ref()
                 == Some(&MetaSubmissionValue {
-                    value: value.clone(),
+                    value: desired_value.clone(),
                     salt,
                 })
             {
                 to_delete.push(key);
-            } else if Self::get_consensus(dbtx, key).await.as_ref() == Some(&value) {
-                // When a value is new and equal to current consensus, it should both: be
-                // submitted so it can clear/cancel previous submissions, and
-                // deleted (afterwards).
-                to_delete.push(key);
-                to_submit.push(MetaConsensusItem { key, value, salt });
+            } else if consensus_value.as_ref() == Some(&desired_value) {
+                if consensus_submission_value.is_none() {
+                    // our desired value cleared the submission (as it is equal the consensus)
+                    // so we can stop proposing it
+                    to_delete.push(key);
```

Comment:
There is a delay between us proposing something and it being accepted, so just because `consensus_submission_value` is `None` now doesn't mean it won't become `Some(_)` at a later point without us sending another proposal. I'm not sure if this is mitigated by the guarantee that earlier proposals will always precede later proposals in consensus output. 

Isn't there any versioning in the votes, so that old votes don't matter anymore anyway? Sounds like a much easier solution. What is that `salt` about? Maybe it can be (ab)used for that?

Reply:
> There is a delay between us proposing something and it being accepted, so just because consensus_submission_value is None now doesn't mean it won't become Some(_) at a later point without us sending another proposal.

Hmmmmm...

For this to happen, the current consensus would had to be some `C`, our current submission `None` (meaning `= C`) and then desired value requested to be set to `X != C` and submitted. While it's propagating, desired value set to `C` again,  and then the user would need to set some other key to something else, because only then we actually act on that `to_delete` and delete it.


Technically still a race that could lead to a request fail to clear that `X`. In practice it's even less probable than it was before, and just as before - probably doesn't matter.

> What is that salt about? Maybe it can be (ab)used for that?

The problem is that the submission equal to consensus is just deleted, which includes the salt, so there's nothing to compare with.


Ehh... I think, since I can really make it technically correct, and don't want to deal with changing the parts that are actually consensus-sentive, I'm just going to stop deleting own desired key at all. Just not worth it, tryint to be too clever by half about it.

Reply:
@elsirion will try https://github.com/fedimint/fedimint/pull/4836 instead